### PR TITLE
Mail: Refactor 0030825: Replace Mail Mandaten

### PR DESCRIPTION
**Pull request by fhelfer**
This fixes: https://mantis.ilias.de/view.php?id=30825

### CODE REVIEW

File: Services/Mail/README.md [hunk 1]
```diff
@@ -632,7 +632,7 @@ is sent:
 * \[IF_NO_PASSWORD\]...\[/IF_NO_PASSWORD\]: This text block is only included, if the new user account has been created without a password.
 * \[ADMIN_MAIL\]: Email address of Administrator
 * \[ILIAS_URL\]: URL of ILIAS system
-* \[CLIENT_NAME\]: Client Name
+* \[INSTALLATION_NAME\]: Client Name
```
Comment (mjansenDatabay) [inline Services/Mail/README.md]:
```suggestion
* \[INSTALLATION_NAME\]: Installation Name
```

File: lang/ilias_de.lang [hunk 2]
```diff
@@ -10757,8 +10757,8 @@ mail#:#mail_my_courses#:#Meine Kurse
 mail#:#mail_my_groups#:#Meine Gruppen
 mail#:#mail_my_mailing_lists#:#Meine Verteilerlisten
 mail#:#mail_nacc_admin_mail#:#E-Mail-Adresse des Administrators
-mail#:#mail_nacc_client_desc#:#Beschreibung des Mandanten
-mail#:#mail_nacc_client_name#:#Name des Mandanten
+mail#:#mail_nacc_installation_desc#:#Beschreibung des Mandanten
+mail#:#mail_nacc_installation_name#:#Name des Mandanten
```
Comment (mjansenDatabay) [inline lang/ilias_de.lang]:
```suggestion
mail#:#mail_nacc_installation_desc#:#Beschreibung der Installation
mail#:#mail_nacc_installation_name#:#Name der Installation
```

File: lang/ilias_en.lang [hunk 3]
```diff
@@ -10712,8 +10712,8 @@ mail#:#mail_my_courses#:#My Courses
 mail#:#mail_my_groups#:#My Groups
 mail#:#mail_my_mailing_lists#:#My Mailing Lists
 mail#:#mail_nacc_admin_mail#:#Mail address of Administrator
-mail#:#mail_nacc_client_desc#:#Client Description
-mail#:#mail_nacc_client_name#:#Client Name
+mail#:#mail_nacc_installation_desc#:#Client Description
+mail#:#mail_nacc_installation_name#:#Client Name
```
Comment (mjansenDatabay) [inline lang/ilias_en.lang]:
```suggestion
mail#:#mail_nacc_installation_desc#:#Installation Description
mail#:#mail_nacc_installation_name#:#Installation Name
```

File: Services/Mail/classes/Setup/class.ilMailDatabaseUpdateSteps.php [hunk 4]
```diff
@@ -31,4 +31,40 @@ public function step_1() : void
...
+                    [
+                        'tpl_id' => (int) $row['tpl_id'],
+                    ],
+                    [
+                        'm_message' => str_replace('[CLIENT_NANE]', '[INSTALLATION_NAME]', $row['m_message'])
+                    ]
+                );
+            }
+        }
+    }
```
Comment (mjansenDatabay) [inline Services/Mail/classes/Setup/class.ilMailDatabaseUpdateSteps.php:83]: IMO we need a further step to replace the content in the `body` field of table `mail_template`.

**Review by mjansenDatabay — CHANGES_REQUESTED** [review]
Thx @fhelfer , LGTM in general. I only have some minor remarks.

We don't need this code anymore: https://github.com/ILIAS-eLearning/ILIAS/pull/4812/files#diff-6347dce422a6e5cf515d769bee07abc81edca22ff56a565019e9bd058f4f86c8L1427

File: Services/Mail/classes/Setup/class.ilMailDatabaseUpdateSteps.php [hunk 5]
```diff
@@ -31,4 +31,59 @@ public function step_1() : void
...
+            );
+        }
+    }
+
+    public function step_3() : void
+    {
+        $result = $this->db->query("SELECT tpl_id, m_message FROM mail_man_tpl WHERE m_message LIKE '%[CLIENT_NAME]%'");
+        while ($row = $this->db->fetchAssoc($result)) {
+            if (isset($row['m_message'], $row['tpl_id']) && $row['m_message'] !== null && $row['tpl_id'] !== null) {
+                $this->db->replace(
```
Comment (mjansenDatabay) [inline Services/Mail/classes/Setup/class.ilMailDatabaseUpdateSteps.php]: Please don't use `REPLACE`.

> Values for all columns are taken from the values specified in the [REPLACE](https://dev.mysql.com/doc/refman/8.0/en/replace.html) statement. Any missing columns are set to their **default** values, just as happens for [INSERT](https://dev.mysql.com/doc/refman/8.0/en/insert.html). You cannot refer to values from the current row and use them in the new row.

File: Services/Mail/classes/Setup/class.ilMailDatabaseUpdateSteps.php [hunk 6]
```diff
@@ -31,4 +31,59 @@ public function step_1() : void
...
+            }
+        }
+    }
+
+    public function step_4() : void
+    {
+        $result = $this->db->query("SELECT lang, type, body FROM mail_template WHERE body LIKE '%[CLIENT_NAME]%'");
+        while ($row = $this->db->fetchAssoc($result)) {
+            if (isset($row['lang'], $row['type'], $row['body']) && $row['body'] !== null) {
+                $this->db->replace(
```
Comment (mjansenDatabay) [inline Services/Mail/classes/Setup/class.ilMailDatabaseUpdateSteps.php]: Please don't use `REPLACE`.

> Values for all columns are taken from the values specified in the [REPLACE](https://dev.mysql.com/doc/refman/8.0/en/replace.html) statement. Any missing columns are set to their **default** values, just as happens for [INSERT](https://dev.mysql.com/doc/refman/8.0/en/insert.html). You cannot refer to values from the current row and use them in the new row.

File: Services/Mail/classes/Setup/class.ilMailDatabaseUpdateSteps.php [hunk 7]
```diff
@@ -31,4 +31,59 @@ public function step_1() : void
             $this->db->dropTableColumn('mail', 'm_email');
         }
     }
+
+    public function step_2() : void
+    {
+        $result = $this->db->queryF('SELECT value FROM settings WHERE module = %s AND keyword = %s', ['text', 'text'], ['common', 'mail_system_sys_signature']);
+        $row = $this->db->fetchAssoc($result);
+        if (isset($row['value']) && $row['value'] !== null) {
+            $this->db->replace(
```
Comment (mjansenDatabay) [inline Services/Mail/classes/Setup/class.ilMailDatabaseUpdateSteps.php]: Please don't use `REPLACE`.

> Values for all columns are taken from the values specified in the [REPLACE](https://dev.mysql.com/doc/refman/8.0/en/replace.html) statement. Any missing columns are set to their **default** values, just as happens for [INSERT](https://dev.mysql.com/doc/refman/8.0/en/insert.html). You cannot refer to values from the current row and use them in the new row.

File: Services/Mail/classes/Setup/class.ilMailDatabaseUpdateSteps.php [hunk 8]
```diff
@@ -31,4 +31,54 @@ public function step_1() : void
             $this->db->dropTableColumn('mail', 'm_email');
         }
     }
+
+    public function step_2() : void
+    {
+        $result = $this->db->queryF('SELECT value FROM settings WHERE module = %s AND keyword = %s', ['text', 'text'], ['common', 'mail_system_sys_signature']);
+        $row = $this->db->fetchAssoc($result);
+        if (isset($row['value']) && $row['value'] !== null) {
```
Comment (mjansenDatabay) [inline Services/Mail/classes/Setup/class.ilMailDatabaseUpdateSteps.php]:
```suggestion
        if (isset($row['value'])) {
```

If `isset($row['value'])` evaluates to `true`, the value cannot be `null`, so the 2nd check is obsolete.

File: Services/Mail/classes/Setup/class.ilMailDatabaseUpdateSteps.php [hunk 9]
```diff
@@ -31,4 +31,54 @@ public function step_1() : void
...
+                );
+            }
+        }
+    }
+
+    public function step_3() : void
+    {
+        $result = $this->db->query("SELECT tpl_id, m_message FROM mail_man_tpl WHERE m_message LIKE '%[CLIENT_NAME]%'");
+        while ($row = $this->db->fetchAssoc($result)) {
+            if (isset($row['m_message'], $row['tpl_id']) && $row['m_message'] !== null && $row['tpl_id'] !== null) {
```
Comment (mjansenDatabay) [inline Services/Mail/classes/Setup/class.ilMailDatabaseUpdateSteps.php]:
```suggestion
            if (isset($row['m_message'], $row['tpl_id'])) {
```

If `isset($row['m_message'], $row['tpl_id'])'` evaluates to `true`, the values cannot be `null`, so the further checks are obsolete.

File: Services/Mail/classes/Setup/class.ilMailDatabaseUpdateSteps.php [hunk 10]
```diff
@@ -31,4 +31,54 @@ public function step_1() : void
...
+                }
+            }
+        }
+    }
+
+    public function step_4() : void
+    {
+        $result = $this->db->query("SELECT lang, type, body FROM mail_template WHERE body LIKE '%[CLIENT_NAME]%'");
+        while ($row = $this->db->fetchAssoc($result)) {
+            if (isset($row['lang'], $row['type'], $row['body']) && $row['body'] !== null) {
```
Comment (mjansenDatabay) [inline Services/Mail/classes/Setup/class.ilMailDatabaseUpdateSteps.php]:
```suggestion
            if (isset($row['lang'], $row['type'], $row['body'])) {
```

If `isset($row['lang'], $row['type'], $row['body'])` evaluates to `true`, the values cannot be `null`, so the further check is obsolete.

**Review by mjansenDatabay — APPROVED** [review]

**Comment by mjansenDatabay** [comment]
@fhelfer Please rebase this branch based on `trunk`.

**Comment by fhelfer** [comment]
> @fhelfer Please rebase this branch based on `trunk`.

Done

**Comment by mjansenDatabay** [comment]
Sorry @fhelfer . Since we created the `release_8` branch on Friday, we'll have to rebase this on `release_8` (only this PR, not the other open PRs) and switch the PR target branch to `release_8` as well.

**Pull request merged**